web/pgadmin/static/js/tree/pgadmin_tree_save_state.js

```diff
@@ -57,29 +57,34 @@ _.extend(pgBrowser.browserTreeState, {
 
   init: function() {
 
-    const save_tree_state_period = usePreferences.getState().getPreferences('browser', 'browser_tree_state_save_interval');
-
-    if (!_.isUndefined(save_tree_state_period) &&  save_tree_state_period.value !== -1) {
-      // Save the tree state every 30 seconds
-      setInterval(this.save_state, (save_tree_state_period.value) * 1000);
-
-      // Fetch the tree last state while loading the browser tree
-      this.fetch_state.apply(this);
-
-      pgBrowser.Events.on('pgadmin:browser:tree:expand-from-previous-tree-state',
-        this.expand_from_previous_state.bind(this));
-      pgBrowser.Events.on('pgadmin:browser:tree:remove-from-tree-state',
-        this.remove_from_cache.bind(this));
-      pgBrowser.Events.on('pgadmin:browser:tree:update-tree-state',
-        this.update_cache.bind(this));
-    } else if (!_.isUndefined(save_tree_state_period)) {
-      getApiInstance().delete(url_for('settings.reset_tree_state'))
-        .catch(function(error) {
-          console.warn(
-            gettext('Error resetting the tree saved state."'), error);
-        });
-    }
-
+    let saveIntervalId;
+    usePreferences.subscribe((prefStore)=>{
+      // Subscribe to listen for preferences change
+      const save_tree_state_period = prefStore.getPreferences('browser', 'browser_tree_state_save_interval')?.value;
+
+      if (saveIntervalId)
+        clearInterval(saveIntervalId);
+
+      if (!_.isUndefined(save_tree_state_period) &&  save_tree_state_period !== -1) {
+        // Save the tree state at given save_tree_state_period
+        saveIntervalId = setInterval(this.save_state, (save_tree_state_period) * 1000);
+        this.fetch_state.apply(this);
+
+        pgBrowser.Events.on('pgadmin:browser:tree:expand-from-previous-tree-state',
```

Comment:
there is problem here. When the preference will change, a event subscription will be added here and it keep on piling up. It should be done only once, and turned off if the tree state interval is disabled.

Reply:
We can deregister the event in case the tree interval is disabled. I read on some blogs and found that duplicates get ignored if we register an event multiple times with the same parameters. I don't know if we have a different implementation of events.

Reply:
Okay, in our case it is being added multiple times, will fix it.